app/lib/account/backend.dart

```diff
@@ -0,0 +1,180 @@
+          ..created = accessTime
+          ..updated = accessTime;
+        _db.commit(inserts: [user]);
+      } else {
+        // update user if e-mail has been changed
+        if (user.updateEmail(auth.email)) {
+          await _db.withTransaction((tx) async {
+            user = (await _db.lookup<User>([userKey])).single;
+            user.updateEmail(auth.email);
+            user.updated = accessTime;
+            tx.queueMutations(inserts: [user]);
+            await tx.commit();
+          });
+        }
+      }
+      return user;
+    });
+
+    if (user == null) {
+      return null;
+    }
+
+    if (!_pool.isClosed) {
+      // Future not awaited: updating the last accessed field does not need to
+      // block the request serving process.
+      _pool.withResource(() async {
+        _retry.retry(() async {
+          await _db.withTransaction((tx) async {
+            final user = (await _db.lookup<User>([userKey])).single;
+            if (user.lastAccess == null ||
+                user.lastAccess.isBefore(accessTime)) {
+              user.lastAccess = accessTime;
+              tx.queueMutations(inserts: [user]);
+              await tx.commit();
+            } else {
+              await tx.rollback();
+            }
+          });
+        });
+      });
+    }
+
+    return user;
+  }
+}
+
+class AuthResult {
+  final String provider;
+  final String userId;
+  final String email;
+
+  AuthResult(this.provider, this.userId, this.email);
+
+  String get compositeId => '$provider:$userId';
+}
+
+abstract class AuthProvider {
+  String get name;
+
+  Future<AuthResult> tryAuthenticate(String accessToken);
+  Future close();
+}
+
+class GoogleOauth2AuthProvider extends AuthProvider {
+  @override
+  final String name;
+  final String _audience;
+  http.Client _httpClient;
+  oauth2_v2.Oauth2Api _oauthApi;
+
+  GoogleOauth2AuthProvider(this.name, this._audience) {
+    _httpClient = http.Client();
+    _oauthApi = oauth2_v2.Oauth2Api(_httpClient);
+  }
+
+  @override
+  Future<AuthResult> tryAuthenticate(String accessToken) async {
+    if (accessToken == null) {
+      return null;
+    }
+    oauth2_v2.Tokeninfo info;
+    try {
+      info = await _oauthApi.tokeninfo(accessToken: accessToken);
+      if (info == null) {
+        return null;
+      }
+
+      if (info.audience != _audience) {
+        _logger.warning('OAuth2 access attempted with invalid audience, '
+            'for email: "${info.email}", audience: "${info.audience}"');
+        return null;
+      }
+
+      if (info.expiresIn == null ||
+          info.expiresIn <= 0 ||
+          info.verifiedEmail != true ||
+          info.email == null ||
+          info.email.isEmpty) {
+        _logger.warning('OAuth2 token info invalid: ${info.toJson()}');
+        return null;
+      }
+
+      return AuthResult(name, info.userId, info.email);
+    } on oauth2_v2.ApiRequestError catch (e) {
+      _logger.info('Access denied for OAuth2 access token.', e);
```

Comment:
Should this logging include the `accessToken`?

Reply:
I'd rather leave the actual token info out of the logs as much as possible.